### ISSUE TITLE
no-undefined-types reports an undefined type that is defined

### ISSUE THREAD

**Issue by lll000111** [gajus/eslint-plugin-jsdoc#144]
v3.15.1

Even within the same file for the exact same type I get both no error and an error about the type being undefined. All errors of this kind are for types defined with `@typedef` and `@global` in other files.

![Screenshot](https://user-images.githubusercontent.com/25476357/50767948-037a3d00-127f-11e9-8495-ad519cf6824a.png)

**Comment by ibbignerd** [comment]
Could you include your typedef and your relevant eslint config?

**Comment by lll000111** [comment]
```js
/**
 * @typedef {string} SHA256Hash
 */
```

JSDoc rules &mdash; by now I disabled all the rules that caused errors for no good reason though:

```
        'jsdoc/check-examples': 'off',
        'jsdoc/check-param-names': 'error',
        'jsdoc/check-tag-names': 'error',
        'jsdoc/check-types': 'error',
        'jsdoc/newline-after-description': 'off',
        'jsdoc/no-undefined-types': 'off',
        'jsdoc/require-description': 'off',
        'jsdoc/require-description-complete-sentence': 'off',
        'jsdoc/require-example': 'off',
        'jsdoc/require-hyphen-before-param-description': 'warn',
        'jsdoc/require-param': 'warn',
        'jsdoc/require-param-description': 'off',
        'jsdoc/require-param-name': 'error',
        'jsdoc/require-param-type': 'error',
        'jsdoc/require-returns': 'off',
        'jsdoc/require-returns-description': 'off',
        'jsdoc/require-returns-type': 'warn',
        'jsdoc/valid-types': 'off',
```

**Comment by ibbignerd** [comment]
One issue is that `Promise` isn't recognized as a type: `message: 'The type \'Promise\' is undefined.',`

Here's the test I used that passes:
```javascript
    {
      code: `
        /**
         * @global
         * @typedef {string} SHA256Hash
         */

        /**
         * @private
         * @param {SHA256Hash} foo
         */
        function quux(foo) {

        }
      `,
      globals: {
        SHA256Hash: true
      }
    },
```

I think you issue is a little more higher level. ESLint doesn't look like it is interfacing with your IDE to pass the global typedefs. 

I'm not too experienced in this field. Maybe @gajus knows more about this.

**Comment by lll000111** [comment]
I used the IDE for the nice screenshot, it's of course the same on the command line. Your test does not help, most uses of that type in all files, and it is one used *everywhere*, cause no problem. I even chose an example where it is and isn't a problem at the same time.

**Comment by PLASMAchicken** [comment]
By doing `"sourceType": "module"` it fixed the problem for me.

**Comment by lll000111** [comment]
@PLASMAchicken I already have exactly that set in my .eslintrc.js

**Comment by brettz9** [comment]
If the `@typedef`'s are defined in other files, eslint-plugin-jsdoc does not support reading these in: #99. The rule currently only looks within the current file. As a workaround, you could add `/* globals  SHA256Hash */` to each file with such a reference.

I therefore think this can be closed as a dupe of 99. As far as "rules that caused errors for no good reason though", we'd need a more specific (and reduced) test case.

**Comment by brettz9** [comment]
@ibbignerd : Re: `Promise` not being defined, you need `env: {es6: true}` in your config to get the ES6 globals like `Promise` added.

@PLASMAchicken : If you are using Node/commonjs, and not actually using ESM, PR #222 may solve an issue whereby `no-undefined-types` was not catching globals.

**Comment by lll000111** [comment]
Closed as duplicate of https://github.com/gajus/eslint-plugin-jsdoc/issues/99
